src/devices/Gpio/Drivers/VirtualGpioController.cs

```diff
@@ -0,0 +1,279 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Threading.Tasks;
+using System.Threading;
+using System.Collections.Generic;
+using System.Linq;
+using System.Collections.Concurrent;
+using System.Device.Gpio;
+using System;
+using System.Device;
+using System.Reflection;
+using Iot.Device.Gpio.Drivers;
+
+namespace Iot.Device.Gpio
+{
+    /// <summary>
+    /// An implementation of a Virtual <see cref="GpioController"/>.
+    /// </summary>
+    public class VirtualGpioController : GpioController
+    {
+        private readonly ConcurrentDictionary<int, GpioPin> _pins = new ConcurrentDictionary<int, GpioPin>();
+        private readonly ConcurrentDictionary<int, PinEvent> _callbackEvents = new ConcurrentDictionary<int, PinEvent>();
+
+        private record PinEvent(PinEventTypes PinEventTypes, PinChangeEventHandler? Callbacks);
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="VirtualGpioController"/> class.
+        /// </summary>
+        public VirtualGpioController()
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="VirtualGpioController"/> class.
+        /// </summary>
+        /// <param name="numberingScheme">The numbering scheme used to represent pins provided by the controller.</param>
+        public VirtualGpioController(PinNumberingScheme numberingScheme)
+        {
+            // Nothing on purpose, as only logical is suported
+            if (numberingScheme != PinNumberingScheme.Logical)
+            {
+                throw new ArgumentException("Only PinNumberingScheme Logical is supported.");
+            }
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="VirtualGpioController"/> class.
+        /// </summary>
+        public VirtualGpioController(Dictionary<int, GpioPin> pins)
+        {
+            foreach (var pin in pins)
+            {
+                _pins.TryAdd(pin.Key, pin.Value);
+            }
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="VirtualGpioController"/> class.
+        /// </summary>
+        public VirtualGpioController(IEnumerable<GpioPin> pins)
+        {
+            int inc = 0;
+            foreach (var pin in pins)
+            {
+                _pins.TryAdd(inc++, pin);
+            }
+        }
+
+        /// <summary>
+        /// Adds a new pin number with an associated <see cref="GpioPin"/>.
+        /// </summary>
+        /// <param name="pinNumber">The pin number.</param>
+        /// <param name="gpioPin">The <see cref="GpioPin"/> to add.</param>
+        public void Add(int pinNumber, GpioPin gpioPin) => _pins.TryAdd(pinNumber, gpioPin);
+
```

Comment:
Considering the complex numbering of the Beaglebone black discussed recently, I would consider two alternatives to the use of integer for the virtual pin numbering.
- Option 1: A string 
- Option 2: An enum. This could be done by making the controller generic and using Enum as constraint for the generic type (which is new in C# 7.3)

In both cases this could help identifying the pins with any friendly name, which is in use on any board.
Going beyond, you can also define a json map loaded in the project with all the pins getting customized names, [similar to this example](https://github.com/dotnet/iot/blob/main/src/devices/BoardLed/BoardLed.cs).

I love the idea of virtualizing pins, but IMO it should/can be more versatile.

Reply:
I've checked all this and if we want to do it we have to keep in mind few things if we want to have done all to the bottom:

* changing int to string will have an impact on performances as comparing strings and int is slower
* adjustments can be done and it will require to change the GpioPin itself to store the string instead of the pin number otherwise it will complicated much more everything. And indeed will complicated the internal manipulation

Alternative:

* if we don't want to use string at the lower level, the possibility is to generate a random, positive, non already existing pin number from the list
* For this implementation, I will then have to add a new constructor, a new list and new methods using the string on top of the int (getting the int will also be possible anyway)

Reply:
With regards to the "alternative" solution, one possibility is to provide the GpioPin an optional `IDictionary<string, int>` to decode the string into the number that makes sense for the board.
As it is optional, if you want pure performance, you can just skip it.
If you want to use it, the `GpioPin` can use the overload that takes a string and try resolving the pin through the dictionary.

Does this make sense to you?

Reply:
The core issue with this is that it can then be used only when in your own code you are calling this. As all bindings takes a GpioController and pin numbers.
So the solution with alternative is the one compatible with everything in that case. It will indeed require a dictionary and few more functions but ultimately, what is going to be manipulated are just in numbers. So you'll still be able to keep the string in the virtual controller but under the hook and for compatibility reasons, it's going to be pure uint.
As an example on how it can looks like:

```csharp
// Note that we can add couple of other constructors
var virtualGpio = new VirtualGpioController();
virtualGpio.Add("A42", gpioPinA);
virtualGpio.Add("B26", gpioPinB);
var myButtonPin = virtualGpio.GetPinNumber("A42");
var button = New Button(buttonPin: myButtonPin, gpio = virtualGpio);
```

In all cases, I would make those adjustments in a separate PR.